### PR TITLE
Add the --macaddress option to specify the MAC address of the tap interface.

### PR DESCRIPTION
**Pull request by bkoehm**
Signed-off-by: Brian Koehmstedt <bkoehm@gmail.com>

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -490,6 +516,7 @@ static void parse_args(int argc, char *const argv[], struct options *options)
         { "outbound-addr", required_argument, NULL, OUTBOUND_ADDR },
         { "outbound-addr6", required_argument, NULL, OUTBOUND_ADDR6 },
         { "disable-dns", no_argument, NULL, DISABLE_DNS },
+        { "hwaddress", required_argument, NULL, HWADDRESS },
```
Comment (AkihiroSuda) [inline main.c:519]: I'd personally prefer to call it `mac` rather than `hwaddress`, because it is actually software :)

File: hwaddr.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                        "store the whole MAC address.\n");
+        return -1;
+    }
+    return (int)(data_ptr - data);
+}
+
+/*
+ * Convert MAC bytes (data) to a MAC string (hwaddr).
+ * The hwaddr buffer must be large enough to contain the MAC string plus a
+ * null terminator.  If data_len is 6 bytes, the hwaddr buffer must be >= 18
```
Comment (AkihiroSuda) [inline hwaddr.c:39]: data_len should be always 6 bytes, no?

File: main.c [hunk 3]
```diff
@@ -286,6 +295,15 @@ static int parent(int sock, int ready_fd, int exit_fd, const char *api_socket,
         }
     }
 #endif
+    if (cfg->vhwaddress_len > 0) {
+        char mac_address[64];
+        memset(mac_address, 0, sizeof(mac_address));
+        if (hwaddr_data_to_hexstring((unsigned char *)cfg->vhwaddress.sa_data,
+                                     cfg->vhwaddress_len, mac_address,
+                                     sizeof(mac_address)) > 0) {
+            printf("* MAC address:     %s\n", mac_address);
```
Comment (AkihiroSuda) [inline main.c:304]: I don't think we need to have `hwaddr_data_to_hexstring`.

Could be like this:
```c
uint8_t *mac=cfg->vhwaddress.sa_data;
printf(" * MAC address:    %02X:%02X:%02X:%02X:%02X:%02X\n",
  mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
```

File: hwaddr.c [hunk 4]
```diff
@@ -0,0 +1,64 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+#include <stdio.h>
+#include <string.h>
+#include "hwaddr.h"
+
+/*
+ * Convert a MAC string (hwaddr) to bytes (data).
+ * hwaddr must be a null-terminated string in the format of xx:xx:xx:xx:xx:xx.
+ * data_buflen is typically 14, which is the sizeof sockaddr sa_data.
```
Comment (AkihiroSuda) [inline hwaddr.c:9]: always 14, no?

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Thanks, almost LGTM but a couple of nits

File: slirp4netns.1.md [hunk 5]
```diff
@@ -90,6 +90,9 @@ specify outbound interface slirp should bind to (ipv6 traffic only)
 `--disable-dns` (since v1.1.0)
 disable built-in DNS (10.0.2.3 by default)
 
+`--hwaddress` (since v1.1.9)
```
Comment (AkihiroSuda) [inline slirp4netns.1.md:93]: @giuseppe Do you think we have to bump up version to v1.2.0 for a new flag? I think this PR is trivial enough to keep the version to be v1.1.x.

Comment (giuseppe) [reply slirp4netns.1.md:93]: I agree with you, v1.1.x is fine

File: hwaddr.c (hunk 2 above)
Comment (bkoehm) [reply hwaddr.c:39]: I've only ever seen 6 bytes but there is [EUI-64](https://standards.ieee.org/content/dam/ieee-standards/standards/web/documents/tutorials/eui.pdf), which I don't know anything about other than it exists.  Wikipedia says "[The IEEE encourages adoption of the more plentiful EUI-64 for non-Ethernet applications.](https://en.wikipedia.org/wiki/MAC_address)"  The only thing that comes to mind for me is "possibility of future expansion."  But I have no problem with hard-coding 6 bytes on both the encoding and decoding side if that is still your preference.

File: main.c (hunk 1 above)
Comment (bkoehm) [reply main.c:519]: Sure, I will make that change.

Comment (bkoehm) [reply main.c]: This has been done, resolving conversation.

File: main.c (hunk 3 above)
Comment (bkoehm) [reply main.c]: This has been done.  Resolving conversation.

File: hwaddr.c (hunk 4 above)
Comment (bkoehm) [reply hwaddr.c]: Done -- hard-coded now to assume a 6-byte or more buffer.  Resolving conversation.

**Comment by bkoehm** [comment]
@AkihiroSuda I've completed the changes you've requested.  I moved the lone remaining function from hwaddr.c to main.c such that hwaddr.c and hwaddr.h are no longer needed and have been removed.  I also did a force push to my branch so as to maintain only one rebased commit for this pull because I'm making the assumption you want only one commit for this instead of a bunch if the pull is merged.

**Comment by AkihiroSuda** [comment]
Thanks, the design looks good but CI is failing

**Comment by bkoehm** [comment]
I can't tell from the github logs why it's failing in the cloud (it passes locally for me).  All I see are things like this at the end of the log:
```
FAIL tests/test-slirp4netns-seccomp.sh (exit status: 1)
Error: Process completed with exit code 2.
```

**Comment by AkihiroSuda** [comment]
I see several "cannot set MAC address: Invalid argument" lines in https://github.com/rootless-containers/slirp4netns/pull/254/checks?check_run_id=1922683592

**Comment by bkoehm** [comment]
Thank you, that was helpful to point that out and got me thinking it was a memory initialization issue since the failures happened before the macaddress test even ran.  Sure enough, I forgot to initialize vmacaddress_len.  I added a single line to main.c, "cfg->vmacaddress_len = 0;", did a force push and now the tests seem to pass.  Normally I would use valgrind to try and catch things like this but slirp4netns seems to make valgrind crash on my system at the point where setns() is called.

File: main.c [hunk 6]
```diff
@@ -612,6 +637,9 @@ static void parse_args(int argc, char *const argv[], struct options *options)
     if (optarg_outbound_addr6 != NULL) {
         options->outbound_addr6 = strdup(optarg_outbound_addr6);
     }
+    if (optarg_macaddress != NULL) {
+        options->macaddress = strdup(optarg_macaddress);
```
Comment (AkihiroSuda) [inline main.c:641]: Shouldn't this raise an error if !options->do_config_network ?

**Review by AkihiroSuda — APPROVED** [review]

File: main.c [hunk 7]
```diff
@@ -612,6 +638,15 @@ static void parse_args(int argc, char *const argv[], struct options *options)
...
     }
+    if (optarg_macaddress != NULL) {
+        if (!options->do_config_network) {
+            fprintf(stderr, "--macaddr cannot be specified when --configure or "
+                            "-c is not specified\n");
+            goto error;
+        } else {
+            options->macaddress = strdup(optarg_macaddress);
+        }
+    }
```
Comment (bkoehm) [inline main.c:649]: @AkihiroSuda Is this what you had in mind?

Comment (AkihiroSuda) [reply main.c:649]: Yes, thanks

**Pull request merged**